OpenSim/Common/DataTable.h

```diff
@@ -0,0 +1,1142 @@
+ *                                                                            *
+ * Copyright (c) 2005-2015 Stanford University and the Authors                *
+ *                                                                            *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may    *
+ * not use this file except in compliance with the License. You may obtain a  *
+ * copy of the License at http://www.apache.org/licenses/LICENSE-2.0.         *
+ *                                                                            *
+ * Unless required by applicable law or agreed to in writing, software        *
+ * distributed under the License is distributed on an "AS IS" BASIS,          *
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   *
+ * See the License for the specific language governing permissions and        *
+ * limitations under the License.                                             *
+ * -------------------------------------------------------------------------- */
+
+/** \file
+* This file defines the  DataTable class, which is used by OpenSim to provide  *
+* an in-memory container for data access and manipulation.                     *
+*/
+
+#ifndef OPENSIM_DATA_TABLE_H_
+#define OPENSIM_DATA_TABLE_H_
+
+// Standard headers.
+#include <memory>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+#include <string>
+#include <type_traits>
+#include <limits>
+// Non-standard headers.
+#include <SimTKcommon.h>
+
+
+namespace OpenSim {
+  class AbstractDataTable;
+  template<typename ET> class DataTable_;
+
+  using DataTable = DataTable_<SimTK::Real>;
+
+  /// Enum to specify if the the input iterator is traversing data row-wise or
+  /// col-wise.
+  /// clang3.6 crashes if this is turned to a "enum class". Until it is fixed,
+  /// use a pre c++11 enum.
+  enum InputItDir{
+    ROWWISE, 
+    COLWISE
+  };
+
+  /// Bind two DataTables by row.
+  template<typename ET>
+  DataTable_<ET> rbindDataTables(const DataTable_<ET>& dt1, 
+                                 const DataTable_<ET>& dt2);
+
+  /// Bind two DataTables by col.
+  template<typename ET>
+  DataTable_<ET> cbindDataTables(const DataTable_<ET>& dt1, 
+                                 const DataTable_<ET>& dt2);
+
+  // Exceptions.
+  class ColumnDoesNotExist;
+  class ColumnHasLabel;
+  class ColumnHasNoLabel;
+  class ZeroElements;
+  class InvalidEntry;
+  class MetaDataKeyExists;
+  class MetaDataKeyDoesNotExist;
+  class MetaDataTypeMismatch;
+} // namespace OpenSim
+
+
+/**----------------------------------------------------------------------------*
+* AbstractDataTable is the base-class of all DataTable_ allowing storage of    *
+* DataTable_ of different types to be stored in containers. See Datatable_ for *
+* details on the interface.                                                    *
+*-----------------------------------------------------------------------------*/
```

Comment:
Being able to put DataTables into a list is one good reason for having AbstractDataTable and standardizing the errors is also valuable,  but I think there is also a general interface that could be very useful. 

For example access to column labels and asking if it contains a column with a specific label would be useful for then selecting Tables from a list. Access to the underlying element type would also be handy.

I would propose that `MetaData` should be accessible by the `AbstractDataTable` interface as well. Since `MetaData` will be a map of AbstractValues, it will not be templated by any element type and thus will serve any DataTable.

I would eventually put support for column units in AbstractDataTable as well.

Many of these were in the original AbstractDataTable prototype.